### PR TITLE
[GH-404] lowest and highest stat targeting strategies

### LINKED ISSUE #404: Increase items and power ups pickup area
Now it's difficult to pick up items or power-ups; you need to walk over them exactly, and it doesn't feel right

image

### PR DESCRIPTION
**Pull request by ncontinanza**
## Motivation

We need to implement two new targeting strategies, that will target the unit with lowest or highest value of a stat, depending the case.

Closes https://github.com/lambdaclass/afk_gacha_game/issues/404

## Summary of changes

- Implement lowest stat strategy
- Implement highest stat strategy
- Add tests

## How to test it?

1. Start the client in the main branch, and the backend with this branch.
2. Pick one of your units and change its basic skill's targeting strategy to lowest. You should end up with something like the following:
```
"targeting_strategy": {
     "type": {"lowest": "health"},
     "target_allies": false
}
```
3. Run `Champions.Config.import_skill_config` in your iex console to apply the changes.
4. Run a battle in the client. To make things easier, you can only leave the unit you modified in your line up.
   4.1. Remember to change all `Logger.info()` to `IO.inspect()`, so that you can see the complete log in your terminal.
5. Check the logs, and verify that the unit you changed is always attacking the opponent's unit with the lowest health.
6. Repeat to test the case for `highest` strategy.

Also, you can check the tests that were added in this PR

## Checklist
- [X] Tested the changes locally.
- [X] Reviewed the changes on GitHub, line by line.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: apps/champions/test/battle_test.exs [hunk 1]
```diff
@@ -1243,6 +1243,206 @@ defmodule Champions.Test.BattleTest do
       assert "team_2" ==
                Champions.Battle.Simulator.run_battle([self_damage_unit], [target_dummy], maximum_steps: maximum_steps).result
     end
+
+    test "Lowest Health" do
+      # This test pairs a unit with a basic skill that deals damage to the enemy with the lowest health, against two target dummies.
+      # The first target dummy has 10 health points and the second one has 9 health points. The dummy with lowest health can kill the opponent in one hit, so if the lowest health targeting strategy fails, the battle will end in a victory for the team_2.
```
Comment (lotuuu) [inline apps/champions/test/battle_test.exs:1249]: If the target dummy can obliterate the attacker with a skill then it's not a target dummy! let's rename it to something more descriptive

File: apps/game_backend/lib/game_backend/units/skills/mechanics/targeting_strategies/target_strategy.ex [hunk 2]
```diff
@@ -18,14 +18,14 @@ defmodule GameBackend.Units.Skills.Mechanics.TargetingStrategies.Type do
 
   def cast(%{factions: factions}), do: {:ok, %{factions: factions}}
   def cast(%{classes: classes}), do: {:ok, %{classes: classes}}
-  def cast(%{lowest: attribute}), do: {:ok, %{lowest: attribute}}
-  def cast(%{highest: attribute}), do: {:ok, %{highest: attribute}}
+  def cast(%{lowest: attribute}), do: {:ok, %{"lowest" => attribute}}
```
Comment (lotuuu) [inline apps/game_backend/lib/game_backend/units/skills/mechanics/targeting_strategies/target_strategy.ex:21]: Is this necessary?

**Review by lotuuu — CHANGES_REQUESTED** [review]

File: apps/game_backend/lib/game_backend/units/skills/mechanics/targeting_strategies/target_strategy.ex (hunk 2 above)
Comment (ncontinanza) [reply apps/game_backend/lib/game_backend/units/skills/mechanics/targeting_strategies/target_strategy.ex]: I thought it was, but it isn't

**Comment by lotuuu** [comment]
As we discussed, try looking into how hard it would be to randomize the order of units that are tied in `stat`, because right now there is a set precedence over units that are tied, making some units more vulnerable to attacks for no reason

File: apps/champions/lib/champions/battle/simulator.ex [hunk 3]
```diff
@@ -1203,6 +1205,22 @@ defmodule Champions.Battle.Simulator do
     }
   end
 
+  defp sort_units_by_stat(units, stat, desc) do
+    Enum.sort(
+      units,
+      &cond do
+        calculate_unit_stat(&1, String.to_atom(stat)) > calculate_unit_stat(&2, String.to_atom(stat)) ->
```
Comment (lotuuu) [inline apps/champions/lib/champions/battle/simulator.ex:1212]: Let's capture calculate_unit_stat before so we run it twice instead of 4 times

**Review by lotuuu — CHANGES_REQUESTED** [review]

**Review by lotuuu — APPROVED** [review]

File: apps/champions/lib/champions/battle/simulator.ex [hunk 4]
```diff
@@ -621,6 +621,34 @@ defmodule Champions.Battle.Simulator do
     [caster.id]
   end
 
+  defp choose_targets_by_strategy(
+         caster,
+         %{type: %{"lowest" => stat}, target_allies: target_allies, count: count},
+         state
```
Comment (Nico-Sanchez) [inline apps/champions/lib/champions/battle/simulator.ex:627]: I have a few suggestions regarding these 2 new function signatures:
1. I wouldn't extract map keys that will only be used in the body and not for pattern-match use. In this case, I would just match the "type" which is the field that differs in these functions. (See https://hexdocs.pm/elixir/main/code-anti-patterns.html#complex-extractions-in-clauses).
2. I don't like some functions extracting "type" field as a value and others as a nested map key. I think every function should do the same, being just the type a value for type key my preferable option. If you need an additional parameter (stat) just add a new key to the map.

File: apps/champions/lib/champions/battle/simulator.ex [hunk 5]
```diff
@@ -621,6 +621,34 @@ defmodule Champions.Battle.Simulator do
...
+       ) do
+    target_team =
+      Enum.filter(state.units, fn {_id, unit} -> unit.team == caster.team == target_allies end)
+
+    Enum.map(target_team, fn {_id, unit} -> unit end)
+    |> sort_units_by_stat(stat, true)
```
Comment (Nico-Sanchez) [inline apps/champions/lib/champions/battle/simulator.ex:633]: This "true" value seems random. I think we should send an atom (:desc or :asc) and do the 2 separate functions, to sort the units in the desired order.

File: apps/champions/lib/champions/battle/simulator.ex [hunk 6]
```diff
@@ -621,6 +621,34 @@ defmodule Champions.Battle.Simulator do
...
+         %{type: %{"highest" => stat}, target_allies: target_allies, count: count},
+         state
+       ) do
+    target_team =
+      Enum.filter(state.units, fn {_id, unit} -> unit.team == caster.team == target_allies end)
+
+    Enum.map(target_team, fn {_id, unit} -> unit end)
+    |> sort_units_by_stat(stat, false)
+    |> Enum.take(count)
+    |> Enum.map(fn unit -> unit.id end)
```
Comment (Nico-Sanchez) [inline apps/champions/lib/champions/battle/simulator.ex:649]: This whole behavior is the same as above, it just changes the sort order. Can we wrap this up in a function, so each parent function calls it with the desired order to sort?

File: apps/champions/lib/champions/battle/simulator.ex [hunk 7]
```diff
@@ -1069,20 +1097,29 @@ defmodule Champions.Battle.Simulator do
...
-            "random"
+          cond do
+            is_binary(targeting_strategy_type) && targeting_strategy_type in @implemented_targeting_strategies ->
+              targeting_strategy_type
+
+            hd(Map.keys(targeting_strategy_type)) in @implemented_targeting_strategies ->
+              targeting_strategy_type
+
+            true ->
+              "random"
```
Comment (Nico-Sanchez) [inline apps/champions/lib/champions/battle/simulator.ex:1122]: This is getting messy. However, I wrote a suggestion in a previous comment that should help to clean this.

**Review by Nico-Sanchez — CHANGES_REQUESTED** [review]

File: apps/champions/lib/champions/battle/simulator.ex (hunk 4 above)
Comment (ncontinanza) [reply apps/champions/lib/champions/battle/simulator.ex]: As discussed, we'll solve your second comment in a separate issue: https://github.com/lambdaclass/afk_gacha_game/issues/479

File: apps/champions/lib/champions/battle/simulator.ex (hunk 7 above)
Comment (ncontinanza) [reply apps/champions/lib/champions/battle/simulator.ex:1134]: As discussed, we'll solve your this comment in a separate issue: https://github.com/lambdaclass/afk_gacha_game/issues/479

**Review by Nico-Sanchez — APPROVED** [review]

**Comment by lotuuu** [comment]
If we're going to change the design principle we're following, as per @Nico-Sanchez's great suggestion, we should go ahead and follow it in our other `choose_targets_by_strategy` while we're at it! We don't want to introduce heterogeneous code practices in our PRs if we can afford it

File: apps/champions/lib/champions/battle/simulator.ex [hunk 8]
```diff
@@ -621,10 +621,40 @@ defmodule Champions.Battle.Simulator do
     [caster.id]
   end
 
-  defp choose_targets_by_strategy(caster, %{type: "all", target_allies: target_allies}, state),
+  defp choose_targets_by_strategy(
+         caster,
+         %{type: %{"lowest" => stat}} = targeting_strategy,
+         state
+       ) do
+    filter_and_choose_units_by_stat(
```
Comment (lotuuu) [inline apps/champions/lib/champions/battle/simulator.ex:629]:
```suggestion
  defp choose_targets_by_strategy(caster, %{type: %{"lowest" => stat}} = targeting_strategy, state) do
```

File: apps/champions/lib/champions/battle/simulator.ex [hunk 9]
```diff
@@ -621,10 +621,40 @@ defmodule Champions.Battle.Simulator do
...
+      :desc
+    )
+  end
+
+  defp choose_targets_by_strategy(
+         caster,
+         %{type: %{"highest" => stat}} = targeting_strategy,
+         state
+       ) do
+    filter_and_choose_units_by_stat(
```
Comment (lotuuu) [inline apps/champions/lib/champions/battle/simulator.ex:644]:
```suggestion
  defp choose_targets_by_strategy(caster, %{type: %{"highest" => stat}} = targeting_strategy, state) do
```

**Review by lotuuu — CHANGES_REQUESTED** [review]
some formatting shenanigans :P

File: apps/champions/lib/champions/battle/simulator.ex [hunk 10]
```diff
@@ -636,6 +666,16 @@ defmodule Champions.Battle.Simulator do
     Enum.take(sorted_units, amount)
   end
 
+  defp filter_and_choose_units_by_stat(units, stat, count, caster, target_allies, order) do
```
Comment (lotuuu) [inline apps/champions/lib/champions/battle/simulator.ex:669]: I'd rename this to `choose_units_by_stat_and_team`, the "filter" word seems confusing

File: apps/champions/lib/champions/battle/simulator.ex (hunk 8 above)
Comment (lotuuu) [reply apps/champions/lib/champions/battle/simulator.ex:629]: suggestion got suggested wrongly :P you will need to make this change manually

File: apps/champions/lib/champions/battle/simulator.ex (hunk 9 above)
Comment (lotuuu) [reply apps/champions/lib/champions/battle/simulator.ex:644]: suggestion got suggested wrongly :P you will need to make this change manually

File: apps/champions/lib/champions/battle/simulator.ex (hunk 8 above)
Comment (lotuuu) [reply apps/champions/lib/champions/battle/simulator.ex:625]: suggestion got suggested wrongly :P you will need to make this change manually

File: apps/champions/lib/champions/battle/simulator.ex (hunk 9 above)
Comment (ncontinanza) [reply apps/champions/lib/champions/battle/simulator.ex:640]: lol I hate when that happens

**Review by lotuuu — CHANGES_REQUESTED** [review]
Last one!

File: apps/champions/lib/champions/battle/simulator.ex (hunk 9 above)
Comment (lotuuu) [reply apps/champions/lib/champions/battle/simulator.ex:640]: yeah it's also uneditable! sucks

File: apps/champions/lib/champions/battle/simulator.ex (hunk 10 above)
Comment (ncontinanza) [reply apps/champions/lib/champions/battle/simulator.ex:665]: I like it!

**Review by lotuuu — APPROVED** [review]

**Review by Nico-Sanchez — APPROVED** [review]

**Pull request merged**